Dashboard Text Cards

joneath opened the pull request:
Implements issue #6360

Commit:
add dashboard text cards

joneath commented on frontend/src/metabase/dashboard/components/DashboardHeader.jsx:
```diff
@@ -106,6 +108,15 @@ export default class DashboardHeader extends Component {
         this.props.onEditingChange(true);
     }
 
+    async onAddTextBox() {
```
I'm guessing this logic should be moved somewhere else. Possibly `dashboard.js`, thoughts?

salsakran replied:
yup, our usual idiom is to wrap those in redux thunkActions.

joneath replied:
How about now that I'm not doing any async behavior? Should this logic be moved? If so, where?

attekei replied:
yeah this could be in `dashboard.js` – it doesn't really change the situation if there is async behavior or not! You seem to already have added `addDashCardToDashboard` and you can just follow the same pattern for this (maybe `addTextDashCardToDashboard` or something like that).

mazameli committed:
type sizes

salsakran reviewed (dismissed):
Text cards should be stored as DashCards.

If they are saved as Cards, then they'll show up in the questions list, which is non-sensical. I'd rather not have a bunch of places filter out cards of "text" type. To date all cards/questions have been directly mapped to queries and my hunch is that there are a lot of places where this is implicitly assumed that will break.

camsaul committed:
Allow card-free Dashcards

camsaul committed:
Make sure cardless dashcards get returned by GET /api/dashboard/:id

Commit:
add support for dashcards without backing cards

joneath commented on frontend/src/metabase/dashboard/dashboard.js:
```diff
@@ -135,6 +135,24 @@ export const addCardToDashboard = function({ dashId, cardId }: { dashId: DashCar
     };
 }
 
+export const addDashCardToDashboard = function({ dashId, dashcardOverrides }: { dashId: DashCardId, dashcardOverrides: { } }) {
```
I've added a dedicated `addDashCardToDashboard` method to be more readable but much of the functionality is shared in `addCardToDashboard`. I can easily overload the `addCardToDashboard` method to handle this extra functionality if that is desired.

mazameli commented:
I took a look at the current state of things with @kdoh this afternoon, and here's a list of some things we noticed:

* The minimum size of cards in terms of dashboard grid units should reduced down to 2x2
* We need to dial in padding/margins a bit more overall. I can take a crack at that.
* Both ordered and unordered lists are missing bullets
* List text should be the same size as paragraph text. I can look at that as well.
* Anchor links should follow our normal link style: bold blue with no underline, then underline on hover
* Images should be liquid, i.e. their width should be set to 100% of their context so that they resize as you resize the card.
* The "remove this question?" confirmation modal shouldn't call text boxes "questions." I suppose the simplest thing would be to just change the modal to always say "card." I can make that change.
* When editing, the text box border should change to blue when it has focus.
* We need to dial in the size and weight of the edit icon vs. the preview icon on text cards. I'll take a look at that.
* The designs show a little bit of static text at the bottom of these cards to let users know they can use markdown syntax.
* Markdown tables aren't rendering
* Code snippets should have some background coloring. Let's rip off what github does, `like this`. 
> Let's also rip off what GH does for block quotes (super low priority, though)

Here's what we were looking at, for context:

![screen shot 2017-11-28 at 3 37 01 pm](https://user-images.githubusercontent.com/2223916/33352629-b6b47f30-d45e-11e7-8310-a63d1c3201eb.png)

joneath commented:
@mazameli thanks for the feedback. The only thing I will push back on is the card's minimum size. I found 4x3 was the smallest that still managed to look and function correctly in edit mode. Here are some screenshots to show my point.
**4x3**
<img width="176" alt="screen shot 2017-11-29 at 16 36 29" src="https://user-images.githubusercontent.com/163979/33406974-23830c10-d524-11e7-819e-e52a1acff724.png">
**2x2**
<img width="92" alt="screen shot 2017-11-29 at 16 37 00" src="https://user-images.githubusercontent.com/163979/33406975-239bf7d4-d524-11e7-8554-27c76cf5bb66.png">

mazameli commented:
Hmm, what's your window size? What I'm trying to figure out is how to enable the use case of using one of this boxes as a standalone heading without it taking up as much vertical real estate as my `Pithy heading` in the screenshot in my previous comment. Do you think we can get away with 4x2?

joneath commented:
Looks like anything below 1200px gets cramped.

880w
<img width="228" alt="screen shot 2017-11-30 at 13 51 41" src="https://user-images.githubusercontent.com/163979/33456979-e693371e-d5d5-11e7-8063-8ffd9cc47c1c.png">

1440w
<img width="398" alt="screen shot 2017-11-30 at 13 52 23" src="https://user-images.githubusercontent.com/163979/33456987-ed64a154-d5d5-11e7-8128-95f786b3ab89.png">

mazameli commented:
K, I feel you. I'll see if I can help things through font sizing, margins, and padding.

Commit:
add text card styling for tables, lists, code blocks, and links

joneath commented:
I've added styling for tables, lists, links, and codeblocks. I've also added the "Styling with Markdown is supported" link. For now I have it pointing to `https://www.metabase.com/docs/latest/`. Also I couldn't find a secondary link style so I assumed underline on hover.

<img width="323" alt="screen shot 2017-11-30 at 14 51 59" src="https://user-images.githubusercontent.com/163979/33459346-7e128524-d5de-11e7-9d6e-c80746b2bcf9.png">

**Hovered**
<img width="361" alt="screen shot 2017-11-30 at 14 51 54" src="https://user-images.githubusercontent.com/163979/33459347-7e2d8c48-d5de-11e7-9c8a-13689e5e19f2.png">

mazameli committed:
persnickety styling tweaks

mazameli commented:
Nice changes. I went in and got anal retentive with sizing and spacing, and I styled blockquotes a bit more as well:

![screen shot 2017-11-30 at 5 07 49 pm](https://user-images.githubusercontent.com/2223916/33463276-754f24de-d5f1-11e7-8185-9a74aca5ca57.png)

Commit:
hide text visualization in visualization select dropdown

joneath commented:
Green build and ready for final reviews.

salsakran commented:
I'm getting save errors on a new dashboard with a single text card when I try to save for the first time after adding the card. 

```
12-05 07:52:08 ERROR metabase.middleware :: PUT /api/dashboard/33/cards 500 (4 ms) (3 DB calls)
{:message "No implementation of method: :can-read? of protocol: #'metabase.models.interface/IObjectPermissions found for class: nil",
 :stacktrace
 ["models.interface$eval24872$fn__24873$G__24861__24882.invoke(interface.clj:93)"
  "models.dashboard$can_read_QMARK_.invokeStatic(dashboard.clj:36)"
  "models.dashboard$can_read_QMARK_.invoke(dashboard.clj:30)"
  "models.interface$eval24872$fn__24903$G__24863__24912.invoke(interface.clj:93)"
  "api.common$write_check.invokeStatic(common.clj:297)"
  "api.common$write_check.invoke(common.clj:290)"
  "api.common$write_check.invokeStatic(common.clj:300)"
  "api.common$write_check.invoke(common.clj:290)"
  "api.dashboard$fn__45351$fn__45354.invoke(dashboard.clj:253)"
  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:248)"
  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:243)"
  "api.dashboard$fn__45351.invokeStatic(dashboard.clj:242)"
  "api.dashboard$fn__45351.invoke(dashboard.clj:242)"
  "middleware$enforce_authentication$fn__42745.invoke(middleware.clj:122)"
  "api.routes$fn__53165.invokeStatic(routes.clj:63)"
  "api.routes$fn__53165.invoke(routes.clj:63)"
  "routes$fn__53253$fn__53254.doInvoke(routes.clj:92)"
  "routes$fn__53253.invokeStatic(routes.clj:88)"
  "routes$fn__53253.invoke(routes.clj:88)"
  "middleware$log_api_call$fn__42858$fn__42860.invoke(middleware.clj:330)"
  "middleware$log_api_call$fn__42858.invoke(middleware.clj:329)"
  "middleware$add_security_headers$fn__42794.invoke(middleware.clj:245)"
  "middleware$bind_current_user$fn__42749.invoke(middleware.clj:142)"
  "middleware$maybe_set_site_url$fn__42798.invoke(middleware.clj:268)"]}
```

The text card has the below:
```
# Hello
[link](https://google.com)
![yay](https://static.boredpanda.com/blog/wp-content/uploads/2016/08/cute-kittens-30-57b30ad41bc90__605.jpg)
```

Going to the activity page or the dashboards list page gives me the below error -
```
12-05 07:54:24 ERROR metabase.middleware :: GET /api/dashboard 500 (15 ms) (14 DB calls)
{:message "No implementation of method: :can-read? of protocol: #'metabase.models.interface/IObjectPermissions found for class: nil",
 :stacktrace
 ["models.interface$eval24872$fn__24873$G__24861__24882.invoke(interface.clj:93)"
  "models.dashboard$can_read_QMARK_.invokeStatic(dashboard.clj:36)"
  "models.dashboard$can_read_QMARK_.invoke(dashboard.clj:30)"
  "models.interface$eval24872$fn__24873$G__24861__24882.invoke(interface.clj:93)"
  "api.dashboard$hydrate_favorites.invokeStatic(dashboard.clj:30)"
  "api.dashboard$hydrate_favorites.invoke(dashboard.clj:27)"
  "api.dashboard$dashboards_list.invokeStatic(dashboard.clj:46)"
  "api.dashboard$dashboards_list.invoke(dashboard.clj:38)"
  "api.dashboard$fn__45167$fn__45168.invoke(dashboard.clj:56)"
  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:248)"
  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:243)"
  "api.dashboard$fn__45167.invokeStatic(dashboard.clj:48)"
  "api.dashboard$fn__45167.invoke(dashboard.clj:48)"
  "middleware$enforce_authentication$fn__42745.invoke(middleware.clj:122)"
  "api.routes$fn__53165.invokeStatic(routes.clj:63)"
  "api.routes$fn__53165.invoke(routes.clj:63)"
  "routes$fn__53253$fn__53254.doInvoke(routes.clj:92)"
  "routes$fn__53253.invokeStatic(routes.clj:88)"
  "routes$fn__53253.invoke(routes.clj:88)"
  "middleware$log_api_call$fn__42858$fn__42860.invoke(middleware.clj:330)"
  "middleware$log_api_call$fn__42858.invoke(middleware.clj:329)"
  "middleware$add_security_headers$fn__42794.invoke(middleware.clj:245)"
  "middleware$bind_current_user$fn__42749.invoke(middleware.clj:142)"
  "middleware$maybe_set_site_url$fn__42798.invoke(middleware.clj:268)"]}
```

mazameli commented:
@kdoh and I took another look at the state of things and here are the remaining to-dos as I see them:

**Things that will require @joneath or @kdoh**

1. Text boxes shouldn't get faded out when a dashboard filter is active. This means we'll either need to A) make them ignore dashboard filters, B) remove the fadeout effect from dashboard cards entirely ([lots of users hate it](https://github.com/metabase/metabase/issues/4220)), or C) change the fade out effect to instead apply a filter icon to filtered cards. @salsakran and I think the long-term solution to this problem is actually adding sections to dashboards so that filters can be applied at a per-section basis. In light of that, I think Option B is our best course of action until sections are shipped.
2. Text boxes' colors should look right in night mode. @kdoh mentioned he might be able to help out on this.
3. I'm back to believing that text boxes should have a minimum height of only 2 grid units. Anything bigger than that looks bad for the heading use case, especially in fullscreen mode. To buy some vertical real estate, I think we should hide the note about markdown being supported at a certain width breakpoint; maybe like 1100px. Re: @joneath's earlier points about things looking too cramped once the screen gets below ~1100px wide, I think we can possibly also reduce some of the padding at that breakpoint, but also, 900px wide is just a silly width to be editing a dashboard and I'm not as concerned about that situation as I am about headings looking terrible on fullscreen dashboards.
4. Right now, when entering dashboard edit mode, the text cards seem to remember if they were in an edit or preview state before. Instead, they should always default to being shown in the edit state.

**Things I'll try to change myself:**

- [ ] Dial in icon alignment with edit text area
- [ ] Point the markdown link to a real URL
- [ ] Fix scrollbar having too much right padding
- [ ] Too much space above headings, so switch them to have margin-bottom instead of margin-top
- [ ] Bump up size of text box `Aa` icon and eyeball icon
- [ ] Fix color of the blue border on text areas
- [ ] See if Lato looks better than Monaco for the composing state of text areas
- [ ] Try a white bg instead of gray bg for the text areas
- [ ] Give text areas rounded corners

mazameli committed:
whole bunch of small style tweaks

mazameli commented:
@kdoh I managed to "fix" the problem of the scrollbar being inset (there was padding around all of the text box content), but the way I did it was dumb because it basically required me to adding right padding or margin to every kind of element that would show up in the text box. I'm having a brain fart here; what's a better way to do this?

mazameli committed:
add a max-width to paragraphs

kdoh commented:
You could do something akin to `.text-card-markdown > *` to target their immediate children which would keep things a little more maintainable. In general though if you're writing new CSS in MB it's a code smell so let me think on this a little more.

camsaul committed:
Don't try to perms-check Cards for cardless DashCards

camsaul committed:
Merge branch 'dashboard-text-cards' of github.com:metabase/metabase into dashboard-text-cards

camsaul committed:
Merge branch 'dashboard-text-cards' of github.com:metabase/metabase into dashboard-text-cards

camsaul committed:
Merge branch 'master' into dashboard-text-cards

camsaul committed:
Fix typo :wrench:

Commit:
Remove fading out filtered cards. Hide text card markdown supported help text on smaller screens.

joneath commented:
@mazameli I've implemented #1, #3, and #4 of your requested changes but I wasn't sure how to turn on night mode (wow there are a lot of hidden states in this app!) so I couldn't tackle #2 myself.

mazameli commented:
@joneath awesome. Hah, yeah, night mode is a bit hidden: you have to go to full screen mode first (top-right button), and then there should be a crescent moon icon that turns night mode on. You can also turn it on with a url parameter but I’m too tired to remember the exact syntax for that atm. Hah

kdoh committed:
night mode text color

joneath commented:
We now have a merge conflict on `resources/migrations/000_migrations.yaml`. I've looked into it and the resolution looks pretty simple but I'm unfamiliar with how the migration system works and worry that it may use the ID to know what migrations to run. If the system does use the ID to determine the migrations to run then I'm worried that just a simple increment in ID may conflict with individual clients (my migration with ID 70 was for text cards but for other clients it may be the conflicting migration). Are my fears unwarranted?

salsakran committed:
Merge branch 'master' into dashboard-text-cards

salsakran commented:
@camsaul can you verify my fixing of the merge conflict in https://github.com/metabase/metabase/pull/6436/commits/216c10e689d1df51c1cbfffb906a7b54b35f4e1e when you get a chance?

salsakran commented:
actually, now that I look at it, they should be in the other order

salsakran committed:
Update 000_migrations.yaml

since migration 70 has already been taken by the metabase_field.database_type addition in master, we should renumber the report_dashboardcard migration to 71.

salsakran commented:
ok, that _should_ work, @joneath

attekei reviewed (commented):
Generally LGTM, I'll still play with the UI a little more tomorrow. 

If we can find a moment when we are both online at the same time, then we can quickly talk through this PR. But that naturally isn't absolutely necessary.

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3987,3 +3987,14 @@ databaseChangeLog:
+  - changeSet:
+      id: 71
+      author: camsaul
+      comment: 'Added 0.27.0'
```
we should fix this comment to say 'Added 0.28.0' now. I can do that

camsaul committed:
Merge branch 'master' into dashboard-text-cards

camsaul committed:
Fix migration comment version number :wrench:

mazameli commented:
From my perspective, barring code review, we could go ahead and merge this. 

There are still two or three small things I'd like to fix, but I don't know that I'd want to hold up merging for them:
1. Inline code and code block backgrounds should change in night mode so that they're legible.
![screen shot 2017-12-13 at 10 54 40 am](https://user-images.githubusercontent.com/2223916/33965026-5d1943cc-e00f-11e7-9dce-40bcc7ee5a5b.png)

2. Inline code should actually be inline; right now it's displaying at block level:
<img width="323" alt="screen shot 2017-12-13 at 2 10 45 pm" src="https://user-images.githubusercontent.com/2223916/33965047-73031b22-e00f-11e7-9ed8-8926ab06eaf4.png">

3. We need to dial in the breakpoint where we start hiding the markdown info string. We should probably also decrease the interior padding of text areas at that breakpoint:
<img width="649" alt="screen shot 2017-12-13 at 2 11 29 pm" src="https://user-images.githubusercontent.com/2223916/33965127-a8477e2c-e00f-11e7-8117-fb748fe36ac7.png">

Commit:
Change inline code markdown to display as inline in text dashcards. Up dashcard breakpoint showing markdown help link

Commit:
move text card creation code to dashboard.js

mazameli committed:
remove help text

mazameli committed:
remove help text styling

mazameli commented:
I made an executive call to remove the markdown help text and put it into the placeholder text of the text areas. It solves a lot of the stupidness we were running into at smaller window widths. 

@kdoh has one last commit he'll be pushing momentarily to fix a small thing with night mode, but after that this looks done from a UI standpoint.

@camsaul, @attekei, @salsakran — have you reviewed the code to your satisfaction?

salsakran reviewed (approved):
Code LGTM

kdoh committed:
night mode code display

kdoh committed:
Merge branch 'dashboard-text-cards' of github.com:metabase/metabase into dashboard-text-cards

camsaul committed:
Minor cleanup :shower:

camsaul reviewed (approved):
:+1: Let's Get This Merged
